daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueGenerators.scala

```diff
@@ -201,34 +189,35 @@ object ValueGenerators {
   def coidValueGen: Gen[ValueContractId] =
     coidGen.map(ValueContractId(_))
 
-  private def valueGen(nesting: Int): Gen[Value] = {
-    Gen.sized(sz => {
-      val newNesting = nesting + 1
-      val nested = List(
-        (sz / 2 + 1, Gen.resize(sz / 5, valueListGen(newNesting))),
-        (sz / 2 + 1, Gen.resize(sz / 5, variantGen(newNesting))),
-        (sz / 2 + 1, Gen.resize(sz / 5, recordGen(newNesting))),
-        (sz / 2 + 1, Gen.resize(sz / 5, valueOptionalGen(newNesting))),
-        (sz / 2 + 1, Gen.resize(sz / 5, valueMapGen(newNesting))),
-      )
-      val flat = List(
-        (sz + 1, dateGen.map(ValueDate)),
-        (sz + 1, Gen.alphaStr.map(ValueText)),
-        (sz + 1, unscaledNumGen.map(ValueNumeric)),
-        (sz + 1, numGen(Decimal.scale).map(ValueNumeric)),
-        (sz + 1, Arbitrary.arbLong.arbitrary.map(ValueInt64)),
-        (sz + 1, Gen.alphaStr.map(ValueText)),
-        (sz + 1, timestampGen.map(ValueTimestamp)),
-        (sz + 1, coidValueGen),
-        (sz + 1, party.map(ValueParty)),
-        (sz + 1, Gen.oneOf(ValueTrue, ValueFalse)),
-      )
-      val all =
-        if (nesting >= MAXIMUM_NESTING) { List() }
-        else { nested } ++
-          flat
-      Gen.frequency(all: _*)
-    })
+  private def nestedGen = Gen.oneOf(
+    valueListGen,
+    variantGen,
+    recordGen,
+    valueOptionalGen,
+    valueMapGen
+  )
+
+  private def flatGen = Gen.oneOf(
+    dateGen.map(ValueDate),
+    Gen.alphaStr.map(ValueText),
+    unscaledNumGen.map(ValueNumeric),
+    numGen(Decimal.scale).map(ValueNumeric),
+    Arbitrary.arbLong.arbitrary.map(ValueInt64),
+    Gen.alphaStr.map(ValueText),
+    timestampGen.map(ValueTimestamp),
+    coidValueGen,
+    party.map(ValueParty),
+    Gen.oneOf(ValueTrue, ValueFalse)
+  )
+
+  def valueGen: Gen[Value] = Gen.lzy {
+    Gen.sized { size =>
+      for {
+        s <- Gen.choose(0, size)
+        value <- if (s < 1) flatGen
+        else Gen.oneOf(flatGen, Gen.resize(size / (s + 1), nestedGen))
```

Comment:
To reduce the use of recursive generators, it can be changed as follows
```
        else Gen.frequency(
          5 -> flatGen,
          1 -> Gen.resize(size / (s + 1), nestedGen)
        )
```

Reply:
We definitely want to reduce that usage; that's why the original code uses `frequency` after all.